File: ts/project/.sdk/tm/rb/test/omni.rb
# The corpus test runner: vendored voxgig_omni driven through its NATIVE
# API (`make_runner(specref, provider)`), presented to the corpus tests in
# the struct-runner shape they already use (`R[:spec]`, `R[:runset]`,
# `R[:runsetflags]`, `R[:client]`). No compat shim is vendored: the
# adapter below IS the whole bridge, per language, per the vendor-tag
# rollout (docs/design/vendor-tag-rollout.md, Decision 4). It is the
# ruby peer of tm/ts/test/omni.ts and tm/py/test/omni.py.
#
# Four local decisions, all required:
#
# 1. SPEC PATH. omni's own spec resolution expects the caller to hand it
#    a usable path. A relative path is absolutized against THIS file's
#    directory (test/), so the existing '../../.sdk/test/test.json'
#    constant keeps working verbatim wherever the suite is run from.
#
# 2. PROVIDER DELEGATION. Corpus-driven contexts get `ctx.client` set to
#    the runner's provider (omni overwrites it on any ctx/args entry). A
#    five-hook provider object HIDES the live SDK from the generated
#    utilities that reach through it - prepare_auth via
#    client.options_map, fetcher via client.mode, feature_add via
#    client.features. So the provider here is a READ-THROUGH view of the
#    live SDK instance: a Hash subclass holding the omni hooks whose
#    method calls fall through to the SDK - the ruby spelling of ts's
#    prototype delegation. (Upstream omni#56 tracks giving the stock
#    provider the same shape.)
#
# 3. UNDEF-ARGUMENT ENTRIES. Entries with no `in`, `args` or `ctx` mean
#    "call the subject the way this port always has". omni's generic rule
#    is `args = [clone(entry.in)]`, one nil - but ruby's retired struct
#    runner passed `[VoxgigStruct::UNDEF]`, the port's own absence
#    sentinel, and subjects like pathify branch on it. `undefargs` below
#    ports the correction from upstream omni's ruby compat shim
#    (voxgig/omni ruby compat/struct.rb), rewriting those entries to an
#    explicit `args` of one sentinel - in memory, for this port only.
#    `wrapsubject` is the same shim's other half: a subject RESULT of
#    UNDEF (getelem's default alt) is translated to omni's ABSENT so the
#    two absence models never meet.
#
# 4. MATCH-VISIBLE CONTEXTS. The SDK's context is a CLASS INSTANCE, not
#    the plain map ts contexts are, and omni's `match` walks the entry's
#    ctx with map access - against a bare instance every `match: {ctx:
#    ...}` assertion would read ABSENT. `ObjView` wraps the live context
#    as a Hash subclass whose map face mirrors the instance's attributes
#    (snake_case attributes answering to the corpus's camelCase keys), so
#    omni's clone-at-match-time takes a faithful snapshot of the
#    POST-EXECUTION context while subjects keep ordinary method access to
#    the live object.
#
# THE VENDORED RUBY PORT LACKS THE omni#54 RUNNER FIXES the TypeScript
# port has at this tag (upstream voxgig/omni#64 landed them for js/go/py
# only). Vendored files are resynced, never edited, so each gap is
# covered HERE instead:
#
# a. `match()` clones its base and the vendored clone()/jsonstr() have no
#    cycle guard. Ruby's live cycles all pass through objects, not plain
#    Hashes, so the raw objects are inert to both walks - the exposure is
#    the map-face views this resolver introduces. ObjView therefore
#    HIDES the cycle edges (`client`/`utility` on every view, `ctx` on
#    error views - runner/SDK bookkeeping the corpus never asserts on)
#    and carries its wrapping ancestry, so a repeated object stays
#    unwrapped and every clone/jsonstr walk terminates.
#
# b. `errify`/`errmessage` collapse non-Exception throwables. Ruby can
#    only raise Exceptions, so that gap cannot fire here; the provider
#    still declares its own `errify` so `match: {err: {code: ...}}` can
#    see an SDK error's code, which the stock {name,message} shape drops.

require 'json'

require_relative 'vendor/omni/voxgig_omni'

module ProjectNameOmni
  TEST_DIR = File.expand_path(__dir__)

  NULLMARK = VoxgigOmni::NULLMARK
  UNDEFMARK = VoxgigOmni::UNDEFMARK
  EXISTSMARK = VoxgigOmni::EXISTSMARK
  ABSENT = VoxgigOmni::ABSENT
  OmniError = VoxgigOmni::OmniError

  # Attribute names the map face of a view does not mirror. client and
  # utility both reach the SDK, whose root context reaches the client
  # again - the cycle omni's guardless clone would follow forever; an
  # error's ctx is the same cycle one hop in (generated make_error
  # attaches the live context to every error it raises). Subjects still
  # reach all three through the method face; the corpus never asserts on
  # any of them.
  VIEW_HIDE = %w[client utility].freeze
  VIEW_HIDE_ERR = %w[client utility ctx].freeze

  # Ruby cannot name an attribute `alias`, so the Spec object spells it
  # `alias_map` - but the corpus speaks the wire name.
  VIEW_ALIAS = { 'alias' => 'alias_map' }.freeze
  VIEW_ALIAS_BACK = { 'alias_map' => 'alias' }.freeze

  module_function

  def camelize(name)
    return VIEW_ALIAS_BACK[name] if VIEW_ALIAS_BACK.key?(name)

    parts = name.split('_')
    parts[0] + parts[1..].map { |p| p.empty? ? p : p[0].upcase + p[1..] }.join
  end

  def underscore(name)
    name.gsub(/([A-Z])/) { '_' + Regexp.last_match(1).downcase }
  end

  def plainval?(val)
    val.nil? || val.is_a?(::String) || val.is_a?(::Numeric) ||
      val.is_a?(::TrueClass) || val.is_a?(::FalseClass) ||
      val.is_a?(::Symbol) || val.is_a?(::Hash) || val.is_a?(::Array) ||
      val.is_a?(::Proc) || val.is_a?(::Method) || val.equal?(ABSENT)
  end

  # A live map view of one object, for omni's match phase.
  #
  # Hash SUBCLASS, because omni's ismap() is an is_a?(Hash) check and its
  # clone() materialises maps via each: cloning a view yields a plain
  # deep snapshot of the object's attributes at that moment - match runs
  # after the subject, so the snapshot carries the mutations the corpus
  # asserts on. Method calls (and assignment) fall through to the live
  # object, so test subjects use the view exactly as they would the
  # object itself.
  #
  # nil-valued attributes are OMITTED from the map face: ruby's "not set"
  # is nil where ts's is undefined, and mirroring nil as a present key
  # would let `__EXISTS__` accept an unset field.
  #
  # `seen` carries the wrapping ancestry so a cyclic reach terminates: a
  # repeated object stays unwrapped, which match reads as ABSENT - no
  # corpus case matches into a cycle.
  class ObjView < ::Hash
    def initialize(target, seen = nil)
      super()
      @view_target = target
      @view_seen = seen || {}.freeze
    end

    def view_target
      @view_target
    end

    # --- method face: the live object ---

    def method_missing(name, *args, &blk)
      t = @view_target
      return t.send(name, *args, &blk) if t.respond_to?(name)

      super
    end

    def respond_to_missing?(name, include_private = false)
      @view_target.respond_to?(name, include_private) || super
    end

    # --- map face: what omni walks ---

    def view_names
      t = @view_target
      hide = t.is_a?(::Exception) ? VIEW_HIDE_ERR : VIEW_HIDE
      t.instance_variables.map { |iv| iv.to_s[1..] }
        .reject { |n| n.start_with?('_') || hide.include?(n) }
        .reject { |n| t.instance_variable_get("@#{n}").nil? }
    end

    def view_wrap(name)
      t = @view_target
      seen = @view_seen
      val = t.instance_variable_get("@#{name}")

      # A view held INSIDE the object graph must continue THIS
      # traversal's ancestry, not restart its own - a restarted chain
      # never terminates on the cycle it re-enters.
      if val.is_a?(ObjView)
        tgt = val.view_target
        return tgt if seen.key?(tgt.object_id)

        return ObjView.new(tgt, seen.merge(t.object_id => true).freeze)
      end
      return val if ProjectNameOmni.plainval?(val) || seen.key?(val.object_id)

      ObjView.new(val, seen.merge(t.object_id => true).freeze)
    end

    def view_attrname(key)
      t = @view_target
      key = key.to_s
      name = VIEW_ALIAS[key] || key
      return name if t.instance_variable_defined?("@#{name}")

      snake = ProjectNameOmni.underscore(name)
      return snake if t.instance_variable_defined?("@#{snake}")

      nil
    end

    def keys
      out = view_names.map { |n| ProjectNameOmni.camelize(n) }
      out << 'message' if @view_target.is_a?(::Exception) && !out.include?('message')
      out
    end

    def each
      keys.each { |k| yield [k, self[k]] }
      self
    end

    def length
      keys.length
    end
    alias size length

    def empty?
      keys.empty?
    end

    def key?(key)
      return true if 'message' == key.to_s && @view_target.is_a?(::Exception)

      !view_attrname(key).nil?
    end
    alias has_key? key?
    alias include? key?
    alias member? key?

    def [](key)
      t = @view_target
      return t.message if 'message' == key.to_s && t.is_a?(::Exception) &&
                          !t.instance_variable_defined?('@message')

      name = view_attrname(key)
      name.nil? ? nil : view_wrap(name)
    end

    def []=(key, val)
      # omni's runner writes `first['client'] = testpack[:client]` into a
      # ctx entry's first argument - route it to the live object, which
      # is where prepare_auth and friends will read it back.
      t = @view_target
      name = view_attrname(key) || ProjectNameOmni.underscore(key.to_s)
      setter = "#{name}="
      if t.respond_to?(setter)
        t.send(setter, val)
      else
        t.instance_variable_set("@#{name}", val)
      end
      val
    end

    def to_s
      "#<ObjView #{@view_target.class}>"
    end
    alias inspect to_s
  end

  # An omni provider that is also the live SDK.
  #
  # omni reads a provider as a mapping (`provider[:subject]`), while
  # corpus code reaches through the runpack's client as an SDK
  # (`client.options_map`, `client.features`). The Hash half holds the
  # hooks; every other method resolves against the live SDK instance,
  # and attribute assignment lands on it too.
  class SdkProvider < ::Hash
    def method_missing(name, *args, &blk)
      sdk = self[:sdk]
      return sdk.send(name, *args, &blk) if !sdk.nil? && sdk.respond_to?(name)

      super
    end

    def respond_to_missing?(name, include_private = false)
      sdk = self[:sdk]
      (!sdk.nil? && sdk.respond_to?(name, include_private)) || super
    end
  end

  # The corpus spells an entity as `{"name" => ...}`; resolve_op wants an
  # object with get_name. Nothing else is read from it.
  class EntityRef
    def initialize(name)
      @name = name
    end

    def get_name
      @name
    end
  end

  # The sdkgen corpus writes contexts as pure JSON, and the ruby context
  # constructor only adopts spec/result/response given as INSTANCES - so
  # the JSON forms are materialised here, exactly as the retired inline
  # runner's make_ctx_from_map did.
  def enrich(ctxmap, ctx)
    spec_map = ctxmap['spec']
    ctx.spec = ProjectNameSpec.new(spec_map) if spec_map.is_a?(::Hash)

    res_map = ctxmap['result']
    if res_map.is_a?(::Hash)
      ctx.result = ProjectNameResult.new(res_map)
      err_map = res_map['err']
      if err_map.is_a?(::Hash) && err_map['message'].is_a?(::String) &&
         !err_map['message'].empty?
        ctx.result.err = ProjectNameError.new('', err_map['message'])
      end
    end

    resp_map = ctxmap['response']
    if resp_map.is_a?(::Hash)
      ctx.response = ProjectNameResponse.new(resp_map)
      unless resp_map['body'].nil?
        body_copy = resp_map['body']
        ctx.response.json_func = -> { body_copy }
      end
      headers = resp_map['headers']
      if headers.is_a?(::Hash)
        lower = {}
        headers.each { |k, v| lower[k.downcase] = v }
        ctx.response.headers = lower
      end
    end

    ctx
  end

  # The omni hooks for an SDK subject - what upstream's compat shim
  # called structprovider, inlined here because this resolver is the one
  # consumer.
  def sdkhooks(sdk)
    utility = sdk.get_utility

    subject = lambda do |name|
      # A subject resolves from the utility (the corpus's camelCase name
      # in ruby's snake_case spelling), or from the struct utilities.
      snake = underscore(name.to_s)
      found = utility.respond_to?(snake) ? utility.send(snake) : nil
      if found.nil? && VoxgigStruct.respond_to?(name)
        found = VoxgigStruct.method(name)
      end
      wrapsubject(found)
    end

    client = lambda do |options|
      # A DEF.client entry becomes another SDK instance - rewrapped with
      # the same delegating shape, not a plain hook object.
      sdkprovider(sdk.class.test(nil, options))
    end

    contextify = lambda do |val|
      ctxmap = val.is_a?(::Hash) ? val.dup : {}
      ent = ctxmap['entity']
      if ent.is_a?(::Hash) && ent['name'].is_a?(::String)
        ctxmap['entity'] = EntityRef.new(ent['name'])
      end
      ctx = utility.make_context.call(ctxmap, nil)
      enrich(ctxmap, ctx)
      ctx.utility = utility
      ctx.options = sdk.options_map if ctx.options.nil?
      ObjView.new(ctx)
    end

    inject = lambda do |options, store|
      VoxgigStruct.respond_to?(:inject) ?
        VoxgigStruct.inject(options, store) : options
    end

    errify = ->(err) { ProjectNameOmni.errify(err) }

    {
      subject: subject,
      client: client,
      contextify: contextify,
      inject: inject,
      errify: errify,
      sdk: sdk,
    }
  end

  # The JSON form of an error, with the error's own attributes carried
  # along - the ruby peer of the vendored python port's errify, which the
  # vendored ruby runner lacks (it keeps only {name,message}, dropping
  # the `code` the corpus asserts refusals by). Used as the provider's
  # errify hook (match.err), and by subjects that answer WITH an error
  # value (makePoint).
  def errify(err)
    return { 'name' => 'Error', 'message' => err.to_s } unless err.is_a?(::Exception)

    out = {}
    err.instance_variables.each do |iv|
      name = iv.to_s[1..]
      next if name.start_with?('_') || VIEW_HIDE_ERR.include?(name)

      val = err.instance_variable_get(iv)
      out[name] = val unless val.nil?
    end
    out['name'] = err.class.name
    out['message'] = err.message
    out
  end

  def sdkprovider(sdk)
    provider = SdkProvider.new
    sdkhooks(sdk).each { |k, v| provider[k] = v }
    provider
  end

  # Replace struct's absence sentinel with omni's, at any depth (from
  # upstream omni's ruby compat shim).
  def absentify(val)
    return ABSENT if val.equal?(VoxgigStruct::UNDEF)
    return val.map { |entry| absentify(entry) } if val.is_a?(::Array)

    if val.instance_of?(::Hash)
      out = {}
      val.each { |key, subval| out[key] = absentify(subval) }
      return out
    end

    val
  end

  # A subject whose result speaks omni's absence model.
  def wrapsubject(subject)
    return subject if subject.nil? || !subject.respond_to?(:call)

    ->(*args) { absentify(subject.call(*args)) }
  end

  # Restore ruby's UNDEF-argument reading for entries carrying none of
  # `in`, `args`, `ctx` (see decision 3 above). Rewritten in memory, for
  # this port only - the corpus on disk is untouched.
  ARGKEYS = %w[in args ctx].freeze

  def noargs?(entry)
    entry.is_a?(::Hash) && (entry.keys & ARGKEYS).empty?
  end

  def undefargs(testspec)
    return testspec unless testspec.is_a?(::Hash) && testspec['set'].is_a?(::Array)
    return testspec unless testspec['set'].any? { |entry| noargs?(entry) }

    patched = testspec['set'].map do |entry|
      if noargs?(entry)
        entry = entry.dup
        entry['args'] = [VoxgigStruct::UNDEF]
      end
      entry
    end

    out = testspec.dup
    out['set'] = patched
    out
  end

  # The suites write flags with string keys (`{ 'null' => false }`); the
  # vendored runner reads them as symbols. Translate (from the upstream
  # compat shim).
  def normflags(flags)
    return {} if flags.nil?
    return flags unless flags.is_a?(::Hash)

    out = {}
    flags.each { |key, val| out[key.is_a?(::String) ? key.to_sym : key] = val }
    out
  end

  # struct's make_runner(testfile, client) signature, backed by vendored
  # omni. Also accepts an already-parsed spec object (omni's own
  # capability), which keeps smoke tests free of fixture files.
  def make_runner(testfile, client)
    specref =
      if testfile.is_a?(::String) && !File.absolute_path?(testfile)
        File.join(TEST_DIR, testfile)
      else
        testfile
      end

    provider = sdkprovider(client)
    runner = VoxgigOmni.make_runner(specref, provider)

    lambda do |name = nil, store = nil|
      runpack = runner.call(name, store.nil? ? {} : store)

      omniflags = runpack[:runsetflags]
      runsetflags = lambda do |testspec, flags = nil, testsubject = nil|
        omniflags.call(undefargs(testspec), normflags(flags),
                       wrapsubject(testsubject))
      end
      runset = ->(testspec, testsubject = nil) { runsetflags.call(testspec, {}, testsubject) }

      {
        spec: runpack[:spec],
        runset: runset,
        runsetflags: runsetflags,
        subject: runpack[:subject],
        client: provider,
      }
    end
  end

  # Convert NULLMARK sentinels back into real nulls.
  def null_modifier(val, key, parent, *rest)
    VoxgigOmni::Runner.nullmodifier(val, key, parent, *rest)
  end
end

File: ts/project/.sdk/tm/py/test/omni.py
# The corpus test runner: vendored voxgig_omni driven through its NATIVE
# API (`makeRunner(specref, provider)`), presented to the corpus tests in
# the struct-runner shape they already use (`R["spec"]`, `R["runset"]`,
# `R["runsetflags"]`, `R["client"]`). No compat shim is vendored: the
# adapter below IS the whole bridge, per language, per the vendor-tag
# rollout (docs/design/vendor-tag-rollout.md, Decision 4). It is the
# python peer of tm/ts/test/omni.ts.
#
# Four local decisions, all required:
#
# 1. SPEC PATH. omni's own spec resolution expects the caller to hand it a
#    usable path. A relative path is absolutized against THIS module's
#    directory (test/), so the existing '../../.sdk/test/test.json'
#    constant keeps working verbatim wherever the suite is run from.
#
# 2. PROVIDER DELEGATION. Corpus-driven contexts get `ctx.client` set to
#    the runner's provider (omni overwrites it on any ctx/args map entry).
#    A five-hook provider object HIDES the live SDK from the generated
#    utilities that reach through it - prepare_auth via
#    client.options_map(), fetcher via client.mode, feature_add appending
#    to client.features. So the provider here is a READ-THROUGH view of
#    the live SDK instance: a dict subclass holding the omni hooks whose
#    attribute access falls through to the SDK, the python spelling of
#    ts's prototype delegation. (Upstream omni#56 tracks giving the stock
#    provider the same shape.)
#
# 3. ZERO-ARGUMENT ENTRIES. Entries with no `in`, `args` or `ctx` mean
#    "call the subject with NO argument". omni's generic rule is
#    `args = [clone(entry.in)]`, which passes one None where python must
#    pass nothing - typify() is 1073741824 where typify(None) is 4194432.
#    `zeroargs` below ports the correction from upstream omni's python
#    compat shim (voxgig/omni python compat/struct.py), rewriting those
#    entries to an explicit empty `args` in memory, for this port only.
#
# 4. MATCH-VISIBLE CONTEXTS. The SDK's context is a CLASS INSTANCE, not
#    the plain map ts contexts are, and omni's `match` walks the entry's
#    ctx with map access - against a bare instance every `match: {ctx:
#    ...}` assertion would read ABSENT. `CtxView` wraps the live context
#    as a dict subclass whose map face mirrors the instance's attributes
#    (snake_case attributes answering to the corpus's camelCase keys), so
#    omni's clone-at-match-time takes a faithful snapshot of the
#    POST-EXECUTION context while subjects keep ordinary attribute access
#    to the live object.

import os

from projectname_sdk.core.spec import ProjectNameSpec
from projectname_sdk.core.result import ProjectNameResult
from projectname_sdk.core.response import ProjectNameResponse
from projectname_sdk.core.error import ProjectNameError
from projectname_sdk.utility.voxgig_struct import voxgig_struct as vs

from test.voxgig_omni import (
    EXISTSMARK,
    NULLMARK,
    UNDEFMARK,
    OmniError,
    islist,
    ismap,
    nullmodifier,
)
from test.voxgig_omni import makeRunner as omni_make_runner


_TEST_DIR = os.path.dirname(os.path.abspath(__file__))


# Attribute names the map face of a context view does not mirror. Both
# reach the client, and the client's root context reaches the client
# again - a cycle omni's guardless clone would follow forever. Subjects
# still reach both through the attribute face; the corpus never matches
# on either.
_CTX_HIDE = ('client', 'utility')


def _camel(name):
    """`status_text` -> `statusText`: the corpus speaks camelCase, the SDK
    is snake_case, and the map face is the corpus-facing side."""
    parts = name.split('_')
    return parts[0] + ''.join(p[:1].upper() + p[1:] for p in parts[1:])


def _is_plain(val):
    return val is None or isinstance(
        val, (str, int, float, bool, dict, list, tuple))


class ObjView(dict):
    """A live map view of one python object, for omni's match phase.

    dict-SUBCLASS, because omni's ismap() is an isinstance check and its
    clone() materialises maps via items(): cloning a view yields a plain
    deep snapshot of the object's public attributes at that moment -
    match runs after the subject, so the snapshot carries the mutations
    the corpus asserts on. Attribute access (and assignment) falls
    through to the live object, so test subjects use the view exactly as
    they would the object itself.

    None-valued attributes are OMITTED from the map face: python's "not
    set" is None where ts's is undefined, and mirroring None as a present
    key would let `__EXISTS__` accept an unset field.

    `seen` carries the wrapping ancestry so a cyclic reach (result.err
    holding the ctx that holds the result) terminates: a repeated object
    stays unwrapped, which match reads as ABSENT - no corpus case
    matches into a cycle.
    """

    def __init__(self, target, seen=None):
        super().__init__()
        object.__setattr__(self, '_target', target)
        object.__setattr__(self, '_seen', seen or frozenset())

    # --- attribute face: the live object ---

    def __getattr__(self, name):
        return getattr(object.__getattribute__(self, '_target'), name)

    def __setattr__(self, name, val):
        setattr(object.__getattribute__(self, '_target'), name, val)

    # --- map face: what omni walks ---

    def _names(self):
        target = object.__getattribute__(self, '_target')
        return [n for n in vars(target)
                if not n.startswith('_') and n not in _CTX_HIDE
                and getattr(target, n, None) is not None]

    def _wrapped(self, name):
        target = object.__getattribute__(self, '_target')
        seen = object.__getattribute__(self, '_seen')
        val = getattr(target, name)
        # A view held INSIDE the object graph (an error that captured the
        # ctx it was raised with holds the view the subject received) must
        # continue THIS traversal's ancestry, not restart its own - a
        # restarted chain never terminates on the cycle it re-enters.
        if isinstance(val, ObjView):
            tgt = object.__getattribute__(val, '_target')
            if id(tgt) in seen:
                return tgt
            return ObjView(tgt, seen | {id(target)})
        if _is_plain(val) or callable(val) or id(val) in seen:
            return val
        return ObjView(val, seen | {id(target)})

    def _attrname(self, key):
        target = object.__getattribute__(self, '_target')
        key = str(key)
        if hasattr(target, key):
            return key
        snake = ''.join(
            '_' + c.lower() if c.isupper() else c for c in key)
        return snake if hasattr(target, snake) else None

    def keys(self):
        return [_camel(n) for n in self._names()]

    def items(self):
        return [(_camel(n), self._wrapped(n)) for n in self._names()]

    def values(self):
        return [self._wrapped(n) for n in self._names()]

    def __iter__(self):
        return iter(self.keys())

    def __len__(self):
        return len(self._names())

    def __contains__(self, key):
        return self._attrname(key) is not None

    def __getitem__(self, key):
        name = self._attrname(key)
        if name is None:
            raise KeyError(key)
        return self._wrapped(name)

    def __setitem__(self, key, val):
        # omni's runner writes `first['client'] = testpack['client']` into
        # a ctx/args entry's first argument - route it to the live object,
        # which is where prepare_auth and friends will read it back.
        setattr(object.__getattribute__(self, '_target'), str(key), val)

    def get(self, key, default=None):
        name = self._attrname(key)
        return default if name is None else self._wrapped(name)


class SdkProvider(dict):
    """An omni provider that is also the live SDK.

    omni reads a provider as a mapping (`provider.get('subject')`), while
    corpus code reaches through the runpack's client as an SDK
    (`client._utility`, `client.options_map()`). The dict half holds the
    hooks; everything else resolves against the live SDK instance, and
    attribute assignment lands on it too.
    """

    def __getattr__(self, name):
        try:
            return self[name]
        except KeyError:
            pass
        try:
            return getattr(self['sdk'], name)
        except KeyError:
            raise AttributeError(name) from None

    def __setattr__(self, name, val):
        setattr(self['sdk'], name, val)


def _snake(name):
    """`makeContext` -> `make_context`: corpus subject names are
    camelCase, the SDK utility is snake_case."""
    return ''.join('_' + c.lower() if c.isupper() else c for c in name)


def _enrich(ctxmap, ctx):
    """The sdkgen corpus writes contexts as pure JSON, and the python
    context constructor only adopts spec/result/response given as
    INSTANCES - so the JSON forms are materialised here, exactly as the
    retired inline runner's _make_ctx_from_map did."""

    spec_map = ctxmap.get('spec')
    if isinstance(spec_map, dict):
        ctx.spec = ProjectNameSpec(spec_map)

    res_map = ctxmap.get('result')
    if isinstance(res_map, dict):
        ctx.result = ProjectNameResult(res_map)
        err_map = res_map.get('err')
        if isinstance(err_map, dict):
            msg = err_map.get('message', '')
            if msg != '':
                ctx.result.err = ProjectNameError('', msg)

    resp_map = ctxmap.get('response')
    if isinstance(resp_map, dict):
        ctx.response = ProjectNameResponse(resp_map)
        body = resp_map.get('body')
        if body is not None:
            body_copy = body
            ctx.response.json_func = lambda: body_copy
        headers = resp_map.get('headers')
        if isinstance(headers, dict):
            ctx.response.headers = {
                k.lower(): v for k, v in headers.items()}

    return ctx


class _EntityRef:
    """The corpus spells an entity as `{"name": ...}`; resolve_op wants
    an object with get_name(). Nothing else is read from it."""

    def __init__(self, name):
        self._name = name

    def get_name(self):
        return self._name


def _sdkhooks(sdk):
    """The omni hooks for an SDK subject - what upstream's compat shim
    called structprovider, inlined here because this resolver is the one
    consumer."""

    utility = sdk._utility
    struct_utils = vs.StructUtility()

    # `client.utility().struct` is how the struct corpus suite reaches the
    # struct utilities (mirrors ts). The python utility object does not
    # carry them natively, so the resolver attaches a StructUtility once.
    if getattr(utility, 'struct', None) is None:
        utility.struct = struct_utils

    def subject(name):
        # A subject resolves from the utility (camelCase or the python
        # spelling), or from the struct utilities.
        found = getattr(utility, name, None)
        if found is None:
            found = getattr(utility, _snake(name), None)
        if found is None:
            found = getattr(struct_utils, name, None)
        return found

    def client(options):
        # A DEF.client entry becomes another SDK instance - rewrapped
        # with the same delegating shape, not a plain hook object.
        return _sdkprovider(type(sdk).test(None, options))

    def contextify(val):
        # The SDK supplies its own context wrapper; the corpus JSON is
        # materialised into it, and the result is served through the
        # match-visible view (decision 4 above).
        ctxmap = dict(val) if isinstance(val, dict) else {}
        ent = ctxmap.get('entity')
        if isinstance(ent, dict) and isinstance(ent.get('name'), str):
            ctxmap['entity'] = _EntityRef(ent['name'])
        ctx = utility.make_context(ctxmap, None)
        _enrich(ctxmap, ctx)
        ctx.utility = utility
        if ctx.options is None:
            ctx.options = sdk.options_map()
        return CtxView(ctx)

    def inject(options, store):
        injector = getattr(struct_utils, 'inject', None)
        if callable(injector):
            return injector(options, store)
        return options

    return {
        'subject': subject,
        'client': client,
        'contextify': contextify,
        'inject': inject,
        'utility': lambda: sdk._utility,
        'tester': lambda options=None: type(sdk).test(None, options),
        'sdk': sdk,
    }


# The context view is just an ObjView rooted at a context; named so call
# sites read as what they hold.
CtxView = ObjView


def _sdkprovider(sdk):
    return SdkProvider(_sdkhooks(sdk))


def zeroargs(testspec):
    """Restore python's zero-argument reading for entries carrying none of
    `in`, `args`, `ctx` (see decision 3 above; ported from upstream omni's
    python compat shim). Rewritten in memory, for this port only - an
    authored `args: []` on disk would break the fixed-arity ports."""
    if not isinstance(testspec, dict) or not islist(testspec.get('set')):
        return testspec

    entries = testspec['set']
    if not any(
        isinstance(e, dict) and not ({'in', 'args', 'ctx'} & set(e.keys()))
        for e in entries
    ):
        return testspec

    patched = []
    for entry in entries:
        if isinstance(entry, dict) and not (
                {'in', 'args', 'ctx'} & set(entry.keys())):
            entry = dict(entry)
            entry['args'] = []
        patched.append(entry)

    testspec = dict(testspec)
    testspec['set'] = patched
    return testspec


def makeRunner(testfile, client):
    """struct's makeRunner(testfile, client) signature, backed by vendored
    omni. Also accepts an already-parsed spec object (omni's own
    capability), which keeps smoke tests free of fixture files."""
    if not isinstance(testfile, str):
        specref = testfile
    elif os.path.isabs(testfile):
        specref = testfile
    else:
        specref = os.path.join(_TEST_DIR, testfile)

    provider = _sdkprovider(client)
    runner = omni_make_runner(specref, provider)

    def structrunner(name=None, store=None):
        runpack = runner(name, store)

        omni_runset = runpack['runset']
        omni_runsetflags = runpack['runsetflags']

        def runset(testspec, testsubject=None):
            return omni_runset(zeroargs(testspec), testsubject)

        def runsetflags(testspec, flags=None, testsubject=None):
            return omni_runsetflags(zeroargs(testspec), flags, testsubject)

        return {
            'spec': runpack['spec'],
            'runset': runset,
            'runsetflags': runsetflags,
            'subject': runpack['subject'],
            'client': provider,
        }

    return structrunner


# struct's flag-modifier name, served from native omni.
nullModifier = nullmodifier


__all__ = [
    'EXISTSMARK',
    'NULLMARK',
    'UNDEFMARK',
    'OmniError',
    'CtxView',
    'SdkProvider',
    'makeRunner',
    'nullModifier',
    'zeroargs',
]
